src/Explain.h

```diff
@@ -32,199 +36,206 @@
+                printError("Usage: output  [<filename>]\n");
             }
-
-            // refresh treePad and allow scrolling
-            if (ncurses && !output) {
-                prefresh(treePad, 0, 0, 0, 0, maxy - 3, maxx - 1);
-                scrollTree(maxx, maxy);
+        } else if (command[0] == "format") {
+            if (command.size() == 2 && command[1] == "json") {
+                ExplainConfig::getExplainConfig().json = true;
+            } else if (command.size() == 2 && command[1] == "proof") {
+                ExplainConfig::getExplainConfig().json = false;
+            } else {
+                printError("Usage: format <json|proof>\n");
             }
+        } else if (command[0] == "exit" || command[0] == "q" || command[0] == "quit") {
+            printPrompt("Exiting explain\n");
+            return false;
+        } else {
+            printError(
+                    "\n----------\n"
+                    "Commands:\n"
+                    "----------\n"
+                    "setdepth <depth>: Set a limit for printed derivation tree height\n"
+                    "explain <relation>(<element1>, <element2>, ...): Prints derivation tree\n"
+                    "explainnegation <relation>(<element1>, <element2>, ...): Enters an interactive\n"
+                    "    interface where the non-existence of a tuple can be explained\n"
+                    "subproof <relation>(<label>): Prints derivation tree for a subproof, label is\n"
+                    "    generated if a derivation tree exceeds height limit\n"
+                    "rule <relation name> <rule number>: Prints a rule\n"
+                    "output <filename>: Write output into a file/disable output\n"
+                    "format <json|proof>: switch format between json and proof-trees\n"
+                    "exit: Exits this interface\n\n");
         }
-        if (ncurses && !output) {
-            endwin();
-        }
+
+        return true;
     }
 
+    /* The main explain call */
+    virtual void explain() = 0;
+
 private:
-    ExplainProvenance& prov;
+    /* Get input */
+    virtual std::string getInput() = 0;
 
-    bool ncurses;
-    bool silentPrompt;
-    WINDOW* treePad = nullptr;
-    WINDOW* queryWindow = nullptr;
-    int maxx = 0, maxy = 0;
+    /* Print a command prompt */
+    virtual void printPrompt(const std::string& prompt) = 0;
 
-    int depthLimit;
+    /* Print a tree */
+    virtual void printTree(std::unique_ptr<TreeNode> tree) = 0;
 
-    /** output stream */
-    std::ostream* output;
+    /* Print any other information */
+    virtual void printInfo(const std::string& info) = 0;
 
-    /** flag for outputting in JSON format */
-    bool json;
+    /* Print an error, such as a wrong command */
+    virtual void printError(const std::string& error) = 0;
 
-    // parse relation, split into relation name and values
+    /* Parse tuple, split into relation name and values */
```

Comment:
Use `/**` to start a doxygen comment for the method. What is str? It seems like it could have a more descriptive name. As it is, you should use `@param` to explain what the parameter does.